Mathlib/CategoryTheory/Abelian/DiagramLemmas/Four.lean

```diff
@@ -30,189 +31,104 @@ We show:
 
 ## Implementation details
 
-To show the mono version, we use pseudoelements. For the epi version, we use a completely different
-arrow-theoretic proof. In theory, it should be sufficient to have one version and the other should
-follow automatically by duality. In practice, mathlib's knowledge about duality isn't quite at the
-point where this is doable easily.
+The diagram of the five lemmas is given by a morphism in the category `ComposableArrows C 4`
+between two objects which satisfy `ComposableArrows.Exact`. Similarly, the two versions of the
+four lemma are stated in terms of the category `ComposableArrows C 3`.
 
-However, one key duality statement about exactness is needed in the proof of the epi version of the
-four lemma: we need to know that exactness of a pair `(f, g)`, which we defined via the map from
-the image of `f` to the kernel of `g`, is the same as "co-exactness", defined via the map from the
-cokernel of `f` to the coimage of `g` (more precisely, we only need the consequence that if `(f, g)`
-is exact, then the factorization of `g` through the cokernel of `f` is monomorphic). Luckily, in the
-case of abelian categories, we have the characterization that `(f, g)` is exact if and only if
-`f ≫ g = 0` and `kernel.ι g ≫ cokernel.π f = 0`, and the latter condition is self dual, so the
-equivalence of exactness and co-exactness follows easily.
+The five lemmas is deduced from the two versions of the four lemma. Both of these versions
+are proved separately. It would be easy to deduce the epi version from the mono version
+using duality, but this would require lengthy API developments for `ComposableArrows` (TODO).
 
 ## Tags
 
 four lemma, five lemma, diagram lemma, diagram chase
 -/
 
 
-open CategoryTheory hiding comp_apply
+namespace CategoryTheory
 
-open CategoryTheory.Abelian.Pseudoelement
+open Category Limits Preadditive
 
-open CategoryTheory.Limits
+namespace Abelian
 
-universe v u
+variable {C : Type*} [Category C] [Abelian C]
 
-variable {V : Type u} [Category.{v} V] [Abelian V]
+open ComposableArrows
 
-attribute [local instance] Preadditive.hasEqualizers_of_hasKernels
+section Four
 
-open Pseudoelement
+variable {R₁ R₂ : ComposableArrows C 3} (hR₁ : R₁.Exact) (hR₂ : R₂.Exact) (φ : R₁ ⟶ R₂)
```

Comment:
The old statements were slightly more general because `hR₁` and `hR₂` contain four exactness statements and both of the four lemmas only require three of them. How hard would it be to state your new versions with the generality of the old versions?

Reply:
I have just implemented this.